### PR TITLE
MAINT: Kill all instances of f2py.compile

### LINKED ISSUE #25122: RFC: Getting rid of `numpy.f2py.compile()`
Here is the implementation:
https://github.com/numpy/numpy/blob/9340fcaeda5aa783a76ab356a58f52b9f76debe3/numpy/f2py/__init__.py#L18-L121

It was lightly amended recently for `meson` use in #25111, however there is really very little justification for its (continued) existence. Some thoughts:

- Things shouldn't be in `__init__`
- It is essentially a wrapper around `subprocess.run`
- Has barely been touched since it was brought in 18 years ago or so (from the `git blame`)
- It [leaks memory](https://github.com/numpy/numpy/blob/9340fcaeda5aa783a76ab356a58f52b9f76debe3/numpy/f2py/tests/test_compile_function.py#L26)

The only slight functionality it provides is that it can work with a block of text (by generating a temporary file). Additionally, tests for `compile` are pretty broken on Windows, and given that it still generates a module file, and the fact that since the `meson` transition users cannot use most of the old `-c` commands, it is really easier to use externally or be explicit about using `subprocess` with environment variables set.

It has always been marked as [experimental and subject to change](https://numpy.org/doc/stable/f2py/usage.html#python-module-numpy-f2py), I would like to therefore remove it ASAP :)

Any user who needs such functionality would be better suited by writing their own thin wrapper over `subprocess` based functionality (or use something like `sh`).

### PR DESCRIPTION
**Pull request by HaoZeke**
Closes #25122 with my favored resolution. There was [no dissent on the mailing list either](https://mail.python.org/archives/list/numpy-discussion@python.org/message/BJB7WHTUU6CC6W3VSJICDGNXUYLKSLM6/). I'm happy to have this float around until `2.0.0` (hence the `versionchanged` in the docs) but it should go, sooner perhaps rather than later.

### CODE REVIEW

File: doc/release/upcoming_changes/25193.compatibility.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+The ``f2py.compile()`` helper has been removed due to a severe restriction in
+functionality after the ``meson`` switch. Users are urged to replace these calls
+with ``subprocess.run`` calls to ``python -m numpy.f2py`` instead with
```
Comment (mattip) [inline doc/release/upcoming_changes/25193.compatibility.rst:3]:
```suggestion
functionality after the ``meson`` switch. Users are urged to replace calls
to ``f2py.compile`` with calls to
``subprocess.run("python", "-m", "numpy.f2py",...`` instead, and to use
```

File: doc/release/upcoming_changes/25193.compatibility.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+environment variables set to interact with ``meson``.
```
Comment (mattip) [inline doc/release/upcoming_changes/25193.compatibility.rst:4]: Hmm. The use of environment variables is problematic. Isn't there any other way to hook into meson?

**Review by mattip — APPROVED** [review]
LGTM. One small nit about the release note, and a question about using environment variables.

File: doc/release/upcoming_changes/25193.compatibility.rst (hunk 1 above)
Comment (HaoZeke) [reply doc/release/upcoming_changes/25193.compatibility.rst:3]: There are [native files](https://mesonbuild.com/Machine-files.html), I've added a note.

File: doc/release/upcoming_changes/25193.compatibility.rst (hunk 2 above)
Comment (HaoZeke) [reply doc/release/upcoming_changes/25193.compatibility.rst]: There are [native files](https://mesonbuild.com/Machine-files.html), I've added a note.

**Comment by rgommers** [comment]
The `test_f2py_init_compile` tests are also very slow and missing the slow test marker, so this will be nice to get in too. @HaoZeke it may be worth running the `f2py` tests with `--durations=25` and adding more `slow` markers. Just the `f2py` tests (default, fast) is taking about 100 seconds for me right now, that is much too long.

File: doc/release/upcoming_changes/25193.compatibility.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+The ``f2py.compile()`` helper has been removed due to a severe restriction in
```
Comment (rgommers) [inline doc/release/upcoming_changes/25193.compatibility.rst:1]: This makes it sound like it's purely due to Meson being worse than distutils, however the linked issue mentioned that it was marked as experimental, leaked memory and was anyway only a thin `subprocess.run` wrapper. So a rephrase here to mention those factors would be nice.

**Review by rgommers — COMMENTED** [review]
This LGTM, after a small rephrase of the release note and resolving the merge conflicts, I think this can go in.

File: doc/release/upcoming_changes/25193.compatibility.rst (hunk 3 above)
Comment (HaoZeke) [reply doc/release/upcoming_changes/25193.compatibility.rst]: Makes sense, reworded and squashed most of the history.

**Review by rgommers — APPROVED** [review]
LGTM now, in it goes. Thanks Rohit!

**Pull request merged**